### ISSUE TITLE
Severe performance issues with `git lfs migrate import` (+24 hours to migrate just a single branch)

### ISSUE THREAD

**Issue by MattMicheletti** [git-lfs/git-lfs#5072]
**Describe the bug**
I have a Git repository that's around a decade old with around 2,000 files that should be in Git LFS (a mix of .jar, .exe, .dll, .pdf, .jpeg, etc. files). I believe `git lfs migrate import` is the correct command to use for this. I have already done `git lfs install` and created a proper `.gitattributes` file. I even weaved it back into the root commit of the `master` branch (i.e. via the following command: `git filter-repo --force --commit-callback "if not commit.parents: commit.file_changes.append(FileChange(b'M', b'.gitattributes', b'$(git hash-object -w .gitattributes)', b'100644'))"`) using `git-filter-repo` and confirmed it worked (as suggested from a prior bug https://github.com/git-lfs/git-lfs/issues/3543#issuecomment-1019633715). I have done several rounds of `git gc`, `git prune`, and `git reflog expire` to ensure the repository is as clean as possible before migration. The repository size (i.e. just the `.git` directory) is just below 5.5 GB before starting `git lfs migrate import`.

Running `git lfs migrate import --fixup --include-ref=refs/heads/master --yes` is **extraordinarily** slow. By slow, I mean that for around 60,000 commits, it takes **over 24 hours** to complete (it took just over 26 hours in my case ... just for the `master` branch alone). That's approximately 1.5 to 1.75 seconds per commit rewrite on average (which is quite slow in the grand scheme of things). I am not even really considering running it for all of our other branches because I do not believe that 24+ hours is the expected behavior for this operation. This must be a bug.

**To Reproduce**
Run `git lfs migrate import --fixup --include-ref=refs/heads/master --yes` on an approximately 5.5. GB repository with around 60,000 commits, around 2,000 binaries of varying types, with a `.gitattributes` file weaved into the root commit.

**Expected behavior**
I would expect that `git lfs migrate import --fixup --include-ref=refs/heads/master --yes` should take _some_ time but absolutely no more than 8 hours at the very maximum and that would be for repositories approaching ungodly sizes of 50+ GB with several 100,000's of commits with a few thousand binaries throughout the history.

Am I just being unreasonable in my expectations here?
Is this normal or within expected parameters for this kind of context?
How would Git LFS' `migrate` functionality be used for moving any given non-trivial Git repository (i.e. even from other SCM systems) to Git LFS without requiring weeks of time?
Is BFG Repo-Cleaner a better approach? Is this a recommended alternative?

**System environment**
OS: Windows 10
Notable Installs: Cygwin, MinGW-w64

Git Config:

```
diff.astextplain.textconv=astextplain
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
http.sslbackend=openssl
http.sslcainfo=C:/Program Files/Git/mingw64/ssl/certs/ca-bundle.crt
core.autocrlf=false
core.fscache=true
core.symlinks=false
core.editor="C:\\Program Files\\Notepad++\\notepad++.exe" -multiInst -notabbar -nosession -noPlugin
pull.rebase=false
credential.helper=manager-core
credential.https://dev.azure.com.usehttppath=true
init.defaultbranch=master
user.name=Abc Xyz
user.email=abc-xyz@test.com
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
difftool.sourcetree.cmd='' "$LOCAL" "$REMOTE"
mergetool.sourcetree.cmd=''
mergetool.sourcetree.trustexitcode=true
```

**Output of `git lfs env`**
Git LFS Environment:

```
git-lfs/3.2.0 (GitHub; windows amd64; go 1.18.2)
git version 2.37.0.windows.1


LocalWorkingDir=
LocalGitDir=
LocalGitStorageDir=
LocalMediaDir=lfs\objects
LocalReferenceDirs=
TempDir=lfs\tmp
ConcurrentTransfers=8
TusTransfers=false
BasicTransfersOnly=false
SkipDownloadErrors=false
FetchRecentAlways=false
FetchRecentRefsDays=7
FetchRecentCommitsDays=0
FetchRecentRefsIncludeRemotes=true
PruneOffsetDays=3
PruneVerifyRemoteAlways=false
PruneRemoteName=origin
LfsStorageDir=lfs
AccessDownload=none
AccessUpload=none
DownloadTransfers=basic,lfs-standalone-file,ssh
UploadTransfers=basic,lfs-standalone-file,ssh
GIT_EXEC_PATH=C:/Program Files/Git/mingw64/libexec/git-core
GIT_LFS_PATH=C:\Program Files\Git LFS
git config filter.lfs.process = "git-lfs filter-process"
git config filter.lfs.smudge = "git-lfs smudge -- %f"
git config filter.lfs.clean = "git-lfs clean -- %f"
```

**Additional context**
There are no remotes involved at the current time. Everything is being done locally, currently. This is part of a migration project from a different SCM system. That is why the repository currently is the way it is.

Please and thank you! :)

**Comment by bk2204** [comment]
Hey,

Migrating a single commit should not take over a second.  Are you perhaps using an antivirus or firewall other than the one built into Windows, or some sort of monitoring software?  Those can make things very slow as they scan every file when it's opened.  Do you see better performance if you try this with WSL?

**Comment by lirantev** [comment]
Hi,
I'm struggling with the same issue myself, but our repo is much bigger (around 49GB in Blob, ~1238 branches and 36449 commits). When I've tried to run the git lfs migrate (from my laptop to our ADS that hosts our repos) on a smaller scale repo the lfs migrate took less than a second for a single file include migrate (~350MB).
Since we have one file which is 1.6 GB (I know this is a big no no but it is what it is for now) I think the performance issue is related to rewriting commits that involves this one big file.
Right now I've executed git lfs migrate command on our huge repo and included a 50 MB file to track with git lfs and the commits rewrites are going faster (at least relative to the previous 1.6GB commits rewrite).
From my 1 day experience migrating our huge repo to GitHub, it seems lfs migrate performance is being affected by lots of things (number of commits, branches, file size).

**Comment by Code7R** [comment]
I was just about to report the same.

I used the dreadful version 3.0.2 which was even much slower. Then 3.1.4 which was a lot better. Now version 3.2 from latest git-for-windows: the feelings are so-so. Sometimes the performance feels okay and sometimes, probably for big files in the commit, it's creepy. Typical commit sizes contain 30-500MB of data. And I see it sometimes processing multiple commits per second, and sometimes stuck on a single commit for 5-10 seconds.

I have checked Process Monitor trace briefly. Looks like it runs single-threaded (in 2022, really?). And it keeps reading content from the PACK file in chunks of 4kB. Doing a syscall for each chunk! This does not make sense, it is even less than regular IO buffer size (8kB TTBOMK). Actually I would read this in 64kB chunks at least.

**Comment by bk2204** [comment]
Hey,

I don't believe the migration code has changed at all since v3.1.4, so the performance should be the same as that version.

However, we are tracking some poor performance with `--above` in #5059.  For now, I would recommend not using `--above` and simply specifying patterns instead.  That will be substantially faster in almost all cases.  We plan to fix that for v3.3.0.

If that's not what you're seeing, we would need to see a testcase of the problem.  We are happy to address issues, but it's very hard to say where the problem is without a reasonable testcase demonstrating the same issue.  I should note that Go programs cannot use memory mapping, but we've verified that the performance of our pure-Go object handling is equivalent to that of calling out to Git.  Also, there are various factors that influence performance, and sometimes the operation can be CPU bound instead of I/O bound, such as when hashing large objects.

I'll also mention that multithreading this is not trivial because Git uses a directed acyclic graph where the object IDs for an object depend on the entire history of the objects that came before it, and obviously there's little benefit to multithreading when you're simply waiting on other work to be done first.

**Comment by Code7R** [comment]
Regarding my previous commit - I was partly wrong. The migrate step itself was finished quite fast. What I am observing here is the initial "git push" after the conversion (pre-push parameter according to process info).

It has been almost stuck for the last couple of hours on the last couple of gigabytes.
Also the syscall pattern has changed. Now it's one of the git.exe child process causing the load. What they do is "interesting", endless sequence of CreateFile, QueryStandardInformationFile, then CreateFileMapping, then git-lfs.exe awakes but it reads ONLY 128 Bytes (probably from mmap-io) and then again CloseFile, CreateFile, ... and the whole repeats again, millions of times.

What is this? A loop of mmap/munmap while fetching with that small buffers?

**Comment by bk2204** [comment]
I would try first doing a `git lfs push --all` and then a `git push --no-verify --force`.  In this case, it's walking the entire history, which is expensive, but `git lfs push --all` will likely be substantially faster.  Regardless, if the migration is fast, then it's not this issue, so that's better discussed in a separate issue.

**Comment by bk2204** [comment]
I've tried migrating really large repositories and I don't see this issue.  The performance is rather reasonable when we've been given testcases or have worked with large repositories.

There are a lot of factors that can affect the performance and all of the things you've mentioned can impact it.  That's why I'm asking for something which demonstrates the problem you're seeing, because there are many dimensions and it's currently very difficult to determine what's the slow part without more context.

**Comment by chrisd8088** [comment]
This may not be helpful, and if so, I apologize.  I just thought I could add a bit of personal perspective.

> Certainly with Git LFS being such an integral part of the Git toolchain there must be a plethora of stress tests, test repositories, and so forth. Do those test suites and test repositories not exhibit such performance degradations?

Git LFS is not actually part of the Git toolchain; it's a separate open-source project maintained by a (very) small group of developers working part-time on it, and one of them (me!) is currently mostly offline due to ongoing family needs.  All of the tests which run are in GitHub Actions and you can review them in our `.github/workflows/ci.yml` [configuration file](https://github.com/git-lfs/git-lfs/blob/72b373eb76a9e855bab7b5d4ae3c55cb45600584/.github/workflows/ci.yml); however, none of them are large performance stress tests or anything like that.

We always welcome contributions to the project and so if you are able to assist us in tracking down the cause of the performance problem by creating a demonstration repository, that would be hugely helpful.  And of course patches, including additions to the test suite, are always most welcome.

> Perhaps it is just a natural flaw of the tool and the technology used to create it? I know nothing about the Go language in regards to pitfalls or limitations other than it's a fairly new language with little done to optimize the compilation process. Could this be the cause of the performance issues noted?

As I've worked on this project and learned more about it, I've come to conceptualize it partly as a wrapper around Git itself.  That's not quite fair, but my general point is that performance issues are not likely to be due to the choice of language, but rather, normal programming concerns like how a particular piece of logic is implemented, or in Git LFS's specific case, how we use the underlying Git commands to manipulate the Git repository.  Sometimes, for instance, we can speed things up by invoking a different set of Git commands or passing different flags, options, and data to Git commands.

This is why it is so crucial to have test-case repositories which demonstrate a performance problem, because we need to understand what aspect of the input data is causing the problem.

> As well, perhaps outlining general performance metrics and benchmarks that should be expected, within reasonable margins of error, could be useful such that we can gauge expected performance parameters given certain input parameters

That sounds like a very useful addition to the [FAQ page](https://github.com/git-lfs/git-lfs/wiki/FAQ) or a dedicated page in the Wiki, and I know we'd welcome contributions from the community to build it up.

> Does `git lfs migrate import` require the user to also use `--exclude-ref` when using `--include-ref`?

No, that shouldn't be necessary.

---

If you're able to help us diagnose the problem further, I would suggest a few potential avenues of exploration.  For instance:

- Running the same Git LFS import on a pure-Linux machine rather than Windows, to see if it's related to the OS.
- Running the same Git LFS import using older versions of Git LFS, say, with the last 2.x version (2.13.3, I believe) to see if there's a regression from earlier behaviour.
- Compiling your own Git LFS version with some debugging statements laced in, or using the Go language's performance profiling enabled, to track down where the code is spending most of its time when running the same Git LFS import.
- Running a variant of the Git LFS import to see if a particular command-line option is causing the extra processing time; for instance, trying without `--fixup` and without the `.gitattributes` in the root commit and instead using `--include` and `--exclude`?

The ultimate goal here is to determine what is distinct about your repository and use case which might introduce a lot of extra processing.  Once we understand that, we can look at whether there are alternatives to the current implementation which improve performance in that case without regressing it in others.

Thank you very much for your help tracking this down, and my apologies again for chiming in from the sidelines.

**Comment by lirantev** [comment]
Hi @chrisd8088,
I'll be happy to assist in anyway possible. regarding the potential avenues of exploration you've mentioned:

- As I don't have access to a pure-linux machine, I wonder if WSL will be sufficient?
- Will try Git LFS 2.13.3 and update with the results
- At least from my side the git lfs command is pretty "simple" - `git lfs migrate --everything --include=*.exe`
As @MattMicheletti pointed out, sharing the repository won't be an option, but I can share the git-sizer output (if it will somehow assist).
Thanks for assistance :)

**Comment by chrisd8088** [comment]
Thanks, @lirantev!  I'm not too familiar with WSL, so I shouldn't speculate too much, but it certainly can't hurt to do the experiment if you have the time.

As for installing an older Windows version of Git LFS, you should be able to get an older binary from our releases list and then, at a minimum, rename the newer binary out of the way and drop the older one in temporarily.  I've done that for testing on Windows, I know.  There may be fancier ways to do this, of course, with the full installer, but for testing it should suffice just to have the older binary run and report the expected older version when you execute `git lfs version`.

**Comment by lirantev** [comment]
Hi,
@chrisd8088 - sorry for the late response, beside the git lfs migrate issue I'm also fighting the almost impossible war of reducing our 12GB repo size to at least 2GB (and hopefully even less).
I've tried to execute git lfs migrate info command using git lfs version 2.13.3 and the performance was great! I've used the command on the same Windows machine I've tried git lfs latest version with the performance issue (so not using WSL or any other change to the OS environment).
I've even tried to execute the same command using the latest git lfs version (I didn't replaced the binaries, to I can use both the git lfs versions, just using the correct 2.13.3 path when using it) and saw the performance issue.
So it seems to me the issue might be related to the git lfs version, and not the repository itself.
If there are any other tests you want me to execute I'll try to do my best to execute them.
Thanks,
Liran.

**Comment by Danielku15** [comment]
Any chance this performance topic is improved? I'm migrating a repository with ~40k commits and just ~ 200 files I want to migrate. I get also only a few commits per second and I have ~250 branches to iterate. After hours I only have a few 1000 commits processed for a single branch. 

As everyone is shouting not to use the BFG Repo Cleaner and "lfs migrate import" is handling the gitattributes a lot cleaner I wonder what the official statement is on migrating large repositories efficiently and reliable. 

I'm on WSL2 on an ext2/ext3 volume (according to `stat -f -c %T .`). 

The Git repository was freshly converted from SVN via Git SVN in ~6hrs.

**Comment by chrisd8088** [comment]
I don't believe there's been a specific improvement or regression in this area for some time.

As has been stated in a few ways in preceding comments in this issue, what would really help would be someone who's experiencing a slow migration process who is either willing to share their repository, or failing that, is able to perform some deep analysis of their own to figure out what kinds of Git repository structures, histories, etc. are causing extra processing in the Git LFS client (or if that's actually the problem at all).

**Comment by Danielku15** [comment]
I understand your need. Unfortunately I think it is often hard to share as mainly companies are doing this transition. Even if we could share the sources, it is a chicken-egg problem. We need to migrate the LFS to upload things to GitHub as that's the whole point. Unfortunately in my case I cannot share anything either. 

Not sure if this helps: 
LFS (Ubuntu Jammy `apt-get install git-lfs`) was running on my repo over night for roughly 16hrs and it only made it to 43% on a single branch. 
I just stumbled over https://github.com/git-lfs/git-lfs/issues/4750 and downgraded to git lfs 3.3.0 via packagecloud.io. It reached 60% on the first branch under 3mins. 

Is there a way of sharing anonymized information about the process which would help here?

**Comment by chrisd8088** [comment]
@Danielku15 -- I'm glad things seem to be improved by changing Git LFS versions.  What version were you using previously?  (You said you downgraded to v3.3.0, but did you perhaps upgrade from v3.0.2, which seems to be what is [provided](https://packages.ubuntu.com/jammy/git-lfs) by Ubuntu Jammy?)  If that's true, then that's a known problem which was indeed fixed in PR #4813.

**Comment by Elhazinforever** [comment]
Are you speaking to me bro or somebody else I even don't know what you
explain and how this world you live in can you be nice person or nice man
I'm in Dubai bro I don't know which  country you are or why all this emails
are you in Dubai are you local or which nationality  and you should check
before send I feel it's criminal or something big wallah bro I really have
alot of problems so if you are blaming me for any case and you are local
please be nice local and don't send anything like this In not bad hundreds
time I proofread I'm straight person how you get my email do you number
also I don't want wake up with  others criminal. You can ask somebody else
in albarsh more people can help thanks if you ard if you in Dubai contact
me if  technology I have friends 😉🤔

**Comment by Danielku15** [comment]
@chrisd8088 As I already upgraded and currently doing the migration I cannot check things right now. I installed a fresh Ubuntu 22.04 LTS on Windows 11 (WSL2) via `apt-get update && apt-get install git git-lfs` assuming this will be the latest and greatest I can get. 😅 

Looking at https://packages.ubuntu.com/jammy/git-lfs it seems to be only 3.0.2. 😣

**Comment by archon810** [comment]
Referencing performance issues in previous comments, I'm trying `git lfs migrate import --fixup --everything` on OpenSUSE right now with version 3.0.2 and it's doing about 1-2 out of 25k commits per second, which is excruciatingly slow.

After compiling the latest version 3.5.1 from source and trying it, it's going about 100x faster. Wow!

**Comment by raphaelsetin** [comment]
### I was able to fix the problem on my end by upgrading Git to the latest version, on Windows 11.
---
Previously , `git lfs migrate import` was stuck for me for several seconds on a repo of almost 2,000 commits.
**NOTE:** I am on WIndows 11, on a normal Terminal, and not in Git Bash.
```shell
$ git lfs version
git-lfs/3.3.0 (GitHub; windows amd64; go 1.19.3; git 77deabdf)
```
---
After upgrading, the migration process for my repo took about 2 minutes.
```shell
$ git update
Warning! `git update` has been deprecated;
Please use `git update-git-for-windows` instead.
... updating ...
```

```shell
$ git lfs version
git-lfs/3.5.1 (GitHub; windows amd64; go 1.21.7; git e237bb3a)
```

**Comment by muhittink** [comment]
Hello everyone,

I'm facing a similiar issue with a long running `git-lfs migrate` command. I have repository with about 60k commits and 20GB, with about 3 Branches, running on linux. The migration process is running now two days and it seems that it's getting slower.
`git-lfs migrate import --yes --everything --include=$FILES`
I was running `[v3.2.0](https://github.com/git-lfs/git-lfs/releases/tag/v3.2.0)` and was very hopeful to see a bugfix in https://github.com/git-lfs/git-lfs/pull/5045 
Because I have observed that replace commits are created under `.git/refs/replace` every 5 or 7 seconds. (now at 35664 commits, at the beginning every 1 - 3 seconds or so). So I updated git-lfs to the latest version but this does not helped me, the migration does not speed up.
So my question is, should `git-lfs migrate import --yes --everything --include=$FILES` create replace refs after  `[v3.3.0](https://github.com/git-lfs/git-lfs/releases/tag/v3.3.0)` ? What I'm missing? We've used this command for about 300, much larger repositories and there wasn't even one that tooks so much time to complete. To my understanding, replace refs aren't necessarily needed by git and should not be written after v3.3.0.

**Comment by muhittink** [comment]
Update: I fixed my problem now! We are running` git-filter-repo` beforehand but without the `--replace-refs delete-no-add` option. After adding this option the migration process is finished in 20 minutes.

**Comment by ramybenaroya** [comment]
@muhittink Could you please elaborate on how `git-filter-repo` is related to `git lfs migrate import`? It's not clear what the order of commands is or how `git-filter-repo` fits into the LFS migration process. Are you using it instead of `git lfs migrate import`, or in addition to it?
